File: scala/E4_1_to_5.scala
// scala E4_1_to_5.scala

case class Some[+A](get: A) extends Option[A]
case object None extends Option[Nothing]

sealed trait Option[+A] {

  def map[B](f: A => B): Option[B] =
    this match {
      case Some(x) => Some(f(x))
      case None => None
    }
  
  def getOrElse[B >: A](default: => B): B =
    this match {
      case Some(x) => x
      case None => default
    }

  def orElse[B >: A](ob: => Option[B]): Option[B] =
    this match {
      case Some(_) => this
      case None => ob
    }

  def flatMap[B](f: A => Option[B]): Option[B] =
    this match {
      case Some(x) => f(x)
      case None => None
    }

  def filter(f: A => Boolean): Option[A] =
    this match {
      case Some(x) if f(x) => this
      case _ => None
    }

}

object Option {

  def mean(xs: Seq[Double]): Option[Double] =
    if (xs.isEmpty) None
    else Some(xs.sum / xs.length)

  def variance(xs: Seq[Double]): Option[Double] =
    mean(xs) flatMap { m => mean(xs map { x => math.pow(x - m, 2) }) }

  def map2[A, B, C](a: Option[A], b: Option[B])(f: (A, B) => C): Option[C] =
    (a, b) match {
      case (Some(x), Some(y)) => Some(f(x, y))
      case _ => None
    }

  def sequence[A](a: List[Option[A]]): Option[List[A]] =
    a match {
      case Nil => Some(Nil)
      case x :: xs =>
        for (
          y <- x;
          ys <- sequence(xs)
        ) yield y :: ys
    }

  def traverse[A, B](a: List[A])(f: A => Option[B]): Option[List[B]] =
    a match {
      case Nil => Some(Nil)
      case x :: xs =>
        for (
          hd <- f(x);
          tl <- traverse(xs)(f)
        ) yield hd :: tl
    }

  def sequence2[A](a: List[Option[A]]): Option[List[A]] = traverse(a)(x => x)

}

File: typescript/src/E4_1_to_5.ts
// yarn run build && node build/src/E4_1_to_5.js

type Option<T> = None | Some<T>;

type None = { tag: 'None' };
const None: Option<never> = { tag: 'None' };
type Some<T> = { tag: 'Some', value: T };
const Some = <T>(value: T): Option<T> => ({ tag: 'Some', value });

// export or else typescript will complain
export const map = <A, B>(f: (x: A) => B, o: Option<A>): Option<B> =>
  o.tag === 'None' ? o : Some(f(o.value));

const getOrElse = <T>(x: T, o: Option<T>): T =>
  o.tag === 'None' ? x : o.value;

const orElse = <T>(a: Option<T>, b: Option<T>) =>
  a.tag === 'Some' ? a : b;

const flatMap = <A, B>(f: (x: A) => Option<B>, o: Option<A>): Option<B> =>
  o.tag === 'None' ? o : f(o.value);

const map2 = <A, B, C>(f: (x: A, y: B) => C, a: Option<A>, b: Option<B>): Option<C> =>
  a.tag === 'Some' && b.tag === 'Some' ? Some(f(a.value, b.value)) : None;

const filter = <T>(f: (x: T) => boolean, a: Option<T>): Option<T> =>
  a.tag === 'Some' && f(a.value) ? a : None;

const mean = (xs: number[]): Option<number> =>
  xs.length === 0 ? None : Some(xs.reduce((x, y) => x + y) / xs.length);

const variance = (xs: number[]): Option<number> =>
  flatMap(m => mean(xs.map(x => Math.pow(x - m, 2))), mean(xs));

const sequence = <T>(xs: Option<T>[]): Option<T[]> =>
  xs.length === 0 ? Some([]) : flatMap(y => flatMap(ys => Some([y].concat(ys)), sequence(xs.slice(1))), xs[0]);

const traverse = <A, B>(f: (x: A) => Option<B>, xs: A[]): Option<B[]> =>
  xs.length === 0 ? Some([]) : flatMap(y => flatMap(ys => Some([y].concat(ys)), traverse(f, xs.slice(1))), f(xs[0]));

const sequence2 = <T>(xs: Option<T>[]): Option<T[]> => traverse(x => x, xs);
